### ISSUE TITLE
Next.js 13 support (server-side components)

### ISSUE THREAD

**Issue by ianjwhite99** [adobe/react-spectrum#3707]
# 🐛 Bug Report 

Unable to use @adobe/react-spectrum with Next.js 13.

While attempting to use react-spectrum with the latest version of Next.js and the new `app/` directory, it appears that the react-spectrum modules are unable to transpile and be used in the latest Next.js version.

## 😯 Current Behavior

Currently, if you install Next.js 13, follow the [Spectrum Server Side](https://react-spectrum.adobe.com/react-spectrum/ssr.html) guide and use the `pages/` directory you are able to successfully use react-spectrum.

However, if you attempt to use the `app/` directory in the latest version of Next you unable to launch the application, even if you follow the Server Side guide and use the `next-transpile-modules` package.

```
error - ./node_modules/@react-spectrum/actiongroup/dist/main.css
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
```

```
error - (sc_server)/node_modules/@react-aria/ssr/dist/main.js (23:94) @ eval
error - TypeError: $parcel$interopDefault(...).createContext is not a function
    at eval (webpack-internal:///(sc_server)/./node_modules/@react-aria/ssr/dist/main.js:26:93)
    at Object.(sc_server)/./node_modules/@react-aria/ssr/dist/main.js
```

## 🔦 Context

Attempting to use @adobe/react-spectrum in a new Next.js 13 project.

## 💻 Code Sample

I've created a repo [spectrum-next13](https://github.com/ianjwhite99/spectrum-next13) that includes an example Next.js 13 project that has react-spectrum installed and is using the `app/` directory. This successfully reproduces the issue mentioned above.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | 3.22.0
| Browser          | Chrome 106.0.5249.119
| Operating System | Mac OS Monterey

**Comment by ianjwhite99** [comment]
Another note, according to the Next.js docs [Styling: CSS-in-JS](https://beta.nextjs.org/docs/styling/css-in-js)

>CSS-in-JS is currently not supported in Server Components. Using CSS-in-JS with newer React features like Server Components and Streaming requires library authors to support the latest version of React, including [concurrent rendering](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react).

**Comment by ianjwhite99** [comment]
Due to Next.js 13 defaulting to server-side components, the temporarily solution while waiting for server-side component support is to use client-side rendering. 

`"use client"` at the top of your component file and render the components client-side.

`app/layout.tsx`
```javascript
"use client";
import { SSRProvider, Provider, defaultTheme } from "@adobe/react-spectrum";

export interface RootLayoutProps {
  children: React.ReactNode;
}

const RootLayout = ({ children }: RootLayoutProps) => {
  return (
    <html>
      <body>
        <SSRProvider>
          <Provider theme={defaultTheme}>
            {children}
          </Provider>
        </SSRProvider>
      </body>
    </html>
  );
};

export default RootLayout;
```

`app/page.tsx`
```javascript
"use client";
import { Button } from "@adobe/react-spectrum";

const RootPage = () => {
  return <Button variant="cta">Hello World</Button>;
};

export default RootPage;
```

**Comment by devongovett** [comment]
Apparently context is completely unsupported in React Server Components. Not sure there is really anything we can do about that, as we rely on context a lot. Appears this will affect a ton of libraries, and the recommended solution is to add `"use client"`: https://github.com/vercel/next.js/issues/41958#issuecomment-1294063309. Hopefully React finds a way to support this.

**Comment by AskYous** [comment]
I get this same error when using [typeorm](https://typeorm.io/).
And `"use client"` is not an option since typeorm uses the filesystem (`require('fs')`).

**Comment by AskYous** [comment]
Oh I apologize. I thought this was nextjs's repository. No it's not related. I wanted to report the issue to nextjs actually.

**Comment by veiico** [comment]
## TL;DR
A solution **might** be removing `<SSRProvider>` usage, and using `useId` to generate ids for all react-aria elements.

## Official way
The ["official" way documented in Next.js](https://beta.nextjs.org/docs/rendering/server-and-client-components#rendering-third-party-context-providers-in-server-components) says that providers can wrap the whole content

It should look something like this:

```jsx
// app/providers.tsx

"use client";

import { SSRProvider } from "react-aria";

export default function Providers({ children }: { children: React.ReactNode }) {
    return <SSRProvider>{children}</SSRProvider>;
}


// app/layout.tsx

import Providers from "~/app/providers";

export default function Layout({ children }: { children: React.ReactNode }) {
    return (
        <html lang="en">
            <body>
                <Providers>
                    {children}
                </Providers>
            </body>
        </html>
    );
}

```

In my case this approach fixes these errors, **BUT** breaks other things...

F.e. popover in this way doesn't get any positioning styles, and is rendered bellow all body content.

## "Solution" (maybe?)
For anyone trying to make this work I've worked around it like this (still testing out, but seems stable enough):
1. Remove SSRProvider completely
2. For every component that has `react-aria` generated props spread-out on them:
2.1 Create a custom SSR id using `useId`
2.2 Add it after the prop-spread

F.e.
```jsx
import { useId } from 'react';

function Button(props) {
  let buttonId = useId(); // create a custom SSR id
  // create as many id's here as there are components using react-aria props

  let ref = useRef();
  let { buttonProps } = useButton(props, ref);
  let { children } = props;

  return (
    <button {...buttonProps} id={buttonId} ref={ref}>
      {children}
    </button>
  );
}
```

I'm using `useId` from React itself, didn't test with react-aria provided one, because the documentation states [here](https://react-spectrum.adobe.com/react-aria/ssr.html#automatic-id-generation):
>If you use React Aria’s [useId](https://react-spectrum.adobe.com/react-aria/useId.html) hook in your own components, SSRProvider will ensure the ids are consistent when server rendered. No additional changes in each component are required to enable SSR support.

As I understood SSRProvider is needed for it to work properly.

At the moment I don't see any problems using React `useId`, but still testing.

If someone sees a problem in this approach - please let us know! 😄

**Comment by snowystinger** [comment]
You're likely going to end up with a lot of relational problems. For example, a menu trigger not announcing the menu it controls via voiceover, a label not being associated with a field, a description not being applied correctly. We have an issue open right now discussing how to go about using reacts useId when it's available.

**Comment by mryechkin** [comment]
Leaving this here for anyone using a library that utilizes React Aria and is getting an error. Here's what I ended up doing:

```jsx
// src/components/Library.js
'use client';

import { Button, Provider } from 'Library';

export { Button, Provider };
```

```jsx
// src/app/layout.js
import { Provider } from 'src/components/Library';

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>
        <Provider>{children}</Provider>
      </body>
    </html>
  );
}
```

This way `RootLayout` can stay a Server Component, and only `Provider`, `Button` etc. are Client Components. Any other components that you'd need to use from the library should be imported and re-exported via `src/components/Library.js` in a similar manner.

**Comment by devongovett** [comment]
I don't think that's quite right. You can pass server components to client components as children, so even if provider was a client component you can still have client components inside.

**Comment by reinierkaper-carewell** [comment]
> I don't think that's quite right. You can pass server components to client components as children, so even if provider was a client component you can still have client components inside.

Are you sure? Because the docs are quite explicit about this (emphasis mine):

>  "use client" sits between server-only and client code. It's placed at the top of a file, above imports, to define the cut-off point where it crosses the boundary from the server-only to the client part. **Once "use client" is defined in a file, all other modules imported into it, including child components, are considered part of the client bundle.**

**Comment by devongovett** [comment]
The key part there is "modules imported into it". So you'd need to put a provider component with "use client" in a separate file from your server code, and pass children into it from a server component instead of rendering the children directly inside the file with "use client".

**Comment by reinierkaper-carewell** [comment]
> The key part there is "modules imported into it". So you'd need to put a provider component with "use client" in a separate file from your server code, and pass children into it from a server component instead of rendering the children directly inside the file with "use client".

Ahh ok, so if you'd have a page that imports a client component and a server component, both rendering children, and would nest the server inside the client component, it would still only ship the client one? That's actually really nice, I think I just confirmed that locally as well. Thanks!

This seems to ship Button and ClientTest, but not ServerTest (even though it's a child of ClientTest, but not imported _in that file_):
![image](https://github.com/adobe/react-spectrum/assets/131210150/0c0c5601-a864-4418-a124-66f807f8ca68)

**Comment by leerob** [comment]
Just want to drop in here and say: client components are not bad and it's totally okay to use them 😄 I understand server components are exciting, but they're also new, so that's why we've focused on making sure you can incrementally adopt things. Using `react-spectrum` with entirely client components in Next.js is totally valid and to be encouraged!

This solution mentioned above should work: https://github.com/adobe/react-spectrum/issues/3707#issuecomment-1556461528
